docs/arcadia-users-group/20231104-testing-concepts/lesson.md

```diff
@@ -0,0 +1,662 @@
+        return 0
+    except TypeError as original_error :
+        msg = "The algebraic mean of an non-numerical list is undefined.\
+               Please provide a list of numbers."
+        raise TypeError(original_error.__str__() + "\n" +  msg)
+```
+
+Now, back in your Jupyter Notebook run the following code:
+
+```
+from mean import *
+
+def test_mean_with_ints():
+    num_list = [1, 2, 3, 4, 5]
+    observed_value = mean(num_list)
+    expected_value = 3
+    assert observed_value == expected_value
+```
+
+The test above:
+
+* sets up the input parameters (the list `[1, 2, 3, 4, 5]`);
+* collects the observed result;
+* declares the expected result (calculated with our human brain);
+* and compares the two with an assertion.
+
+A unit test suite is made up of many tests just like this one.
+A single implemented function may be tested in numerous ways.
+
+In a file called `test_mean.py`, implement the following code:
+
+```
+from mean import *
+
+def test_mean_with_ints():
+    num_list = [1, 2, 3, 4, 5]
+    observed_value = mean(num_list)
+    expected_value = 3
+    assert observed_value == expected_value
+
+def test_mean_with_zero():
+    num_list=[0, 2, 4, 6]
+    observed_value = mean(num_list)
+    expected_value = 3
+    assert observed_value == expected_value
+
+def test_mean_with_double():
+    # This one will fail in Python 2
+    num_list=[1, 2, 3, 4]
+    observed_value = mean(num_list)
+    expected_value = 2.5
+    assert observed_value == expected_value
+
+def test_mean_with_long():
+    big = 100000000
+    observed_value = mean(range(1,big))
+    expected_value = big/2.0
+    assert observed_value == expected_value
+
+def test_mean_with_complex():
+    # given that complex numbers are an unordered field
+    # the arithmetic mean of complex numbers is meaningless
+    num_list = [2 + 3j, 3 + 4j, -32 - 2j]
+    observed_value = mean(num_list)
+    expected_value = NotImplemented
+    assert observed_value == expected_value
+```
+
+Use Jupyter Notebook to import the `test_mean` package and run each test like this:
+
+```
+from test_mean import *
+
+test_mean_with_ints()
+test_mean_with_zero()
+test_mean_with_double()
+test_mean_with_long()
+test_mean_with_complex()  ## Please note that this one might fail. You'll get an error message showing which tests failed
+```
+
+We just wrote and ran five tests for our `mean()` function~
+
+You may have noticed that several of the tests look very similar to each other -- they introduce an input, call `mean()`, and test its output against an expected value. We'll come back to this later, offering a way to write a single test that applies to multiple inputs.
+## Using the test framework `pytest`
+
+We created a suite of tests for our mean function, but it was annoying to run them one at a time.
+It would be a lot better if there were some way to run them all at once, just reporting which tests fail and which succeed.
+
+Thankfully, that exists.
+Recall our tests:
+
+```
```

Comment:
nit: use "```python" to turn on syntax highlighting for longer blobs, for readability.

Reply:
+1 almost mentioned this but felt like I'd spent my nit budget on all those nomenclature changes 😛

Reply:
there is no nit budget! i didn't actually know about syntax highlighting in markdown that way! I'll go through and add it, thanks guys!